[JENKINS-60972] Set 'Dashboard' as first breadcrumb

romenrg opened the pull request:
This can be replaced when a proper Breadcrumbs module is built in the "backend" but since such bigger effort seems to have been hanging for several months (https://github.com/jenkinsci/jenkins/pull/4634) I think this workaround might be a good temporary solution to get the "Home" breadcrumb already in place.

See [JENKINS-60972](https://issues.jenkins-ci.org/browse/JENKINS-60972).

_EDIT: As per the discussion in the comments, we are opting for "Dashboard" instead of "Home", as it is already referenced as "dashboard" in many places in the codebase._

### Screenshot of the result

<img width="1918" alt="Screen Shot 2020-08-27 at 18 26 26" src="https://user-images.githubusercontent.com/1831480/91474984-d682ff80-e892-11ea-9541-71f2e359159e.png">

_EDIT: See updated screenshot in a comment [below](https://github.com/jenkinsci/jenkins/pull/4920#issuecomment-683278880)._

### Proposed changelog entries

* UX: First Breadcrumb (home page) is now named "Dashboard"

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [X] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@fqueiruga @batmat @schottsfired @sladyn98 @oleg-nenashev 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

romenrg committed:
[JENKINS-60972] Use "Dashboard" for 1st breadcrumb (w/ i18n)

This is a squashed commit of 8 original commits:

[JENKINS-60972] Workaround to use "Home" for 1st breadcrumb

This can be replaced when a proper Breadcrumbs module is build in the "backend" but might be a good temporary solution for now

[JENKINS-60972] Using Dashboard instead of Home for first breadcrumb

Dashboard is used in multiple places in the codebase to refer to that page, so seems a better fit. Otherwise, we'd need to rename all those other references to "home"

[JENKINS-60972] Added comment to breadcrumb if for "Dashboard"

(actually used this comment to trigger a rebuild)

[JENKINS-60972] Using choose/when/otherwise + anc.object == app

Better readability and avoiding replacing jobs named Jenkins, for instance

[JENKINS-60972] Adding i18n for breadcrumb

[JENKINS-60972] Simpler cluase comparing with "app" is enough. Tested.

And the extra comparison might lead to unexpected behaviors if there are extra spaces in the displayName in a particular language

[JENKINS-60972] Moved clarifying comment about breadcrumb  to the "choose"

[JENKINS-60972] Using title-case for i18n

daniel-beck reviewed (changes requested):
This page has never been called _Home_ as far as I'm aware, it is always _Dashboard_ (like all those "Back to Dashboard" links in sidepanels). I'm not opposed to "Home" as a term, but then we should change all references to the same page currently labeled "Dashboard" at the same time.

Additionally, this would be less repetitive and less errorprone with `choose`/`when`/`otherwise` elements.

romenrg commented:
@daniel-beck I'm totally OK with using "Dashboard", if that is more widely accepted.
I used "home" as I saw that term proposed in the Jira ticket, but I would say "Dashboard" would be good from a UX perspective as well (definitely better than "Jenkins").
So let's see what others say, but I'm OK with either "Home" or "Dashboard".
Actually, if there is other stuff that needs to be renamed if "Home" is used, then I would opt for "Dashboard" to avoid all the renaming.

michaelneale commented:
I like "home" more, as it is wonderfully generic, when you look at it everyone knows what it means, and "dashboard" is also used to mean various information radiator concepts and full screen views (and is longer). However, if the impact on other things and i18n is non trivial, Dashboard is fine too. 

Anything to make it not say Jenkins Jenkins Jenkins Jenkins which reminds me of "Being John Malkovic" movie.

daniel-beck commented:
> However, if the impact on other things and i18n is non trivial, Dashboard is fine too.

https://github.com/jenkinsci/jenkins/search?q=Dashboard&unscoped_q=Dashboard enjoy

michaelneale commented:
@daniel-beck woah... dashboard it is!

romenrg commented:
Yeah, I agree @michaelneale, let's go with @daniel-beck 's  suggestion of using "Dashboard" to avoid changing all those references. Just made the change and pushed it here.

schottsfired commented:
+1 for Dashboard, sounds good.

daniel-beck reviewed (changes requested):
Should still use choose/when/otherwise to reduce repetitiveness.

Looks like it would rewrite job/folder names from "Jenkins" to "Dashboard" too, as it's not actually limited to `anc.object == app`.

zbynek commented:
Also it would be nice to use [Jelly i18n](https://www.jenkins.io/doc/developer/internationalization/i18n-jelly-views/).

daniel-beck commented on core/src/main/resources/lib/layout/breadcrumbBar.jelly:
```diff
@@ -44,8 +44,14 @@ THE SOFTWARE.
                     <a href="${anc.url}/" class="${h.isModelWithContextMenu(anc.object)?'model-link':null} inside">
-                      ${anc.object.displayName}
+                      <j:if test="${anc.object.displayName == 'Jenkins'}">
+                        Dashboard
```
```suggestion
                        ${%Dashboard}
```

This takes care of i18n.

daniel-beck commented on core/src/main/resources/lib/layout/breadcrumbBar.jelly:
```diff
@@ -44,8 +44,14 @@ THE SOFTWARE.
+                    <!-- Workaround to set desired name to first breadcrumb (instead of Jenkins) -->
                     <a href="${anc.url}/" class="${h.isModelWithContextMenu(anc.object)?'model-link':null} inside">
-                      ${anc.object.displayName}
+                      <j:if test="${anc.object.displayName == 'Jenkins'}">
```
```suggestion
                      <j:if test="${anc.object == app}">
```

Limit to the actual root element. Untested.

romenrg replied:
Applied and tested. Works great!

romenrg commented:
@daniel-beck I had completely missed the second part of your initial comment, sorry. Now I think your and @zbynek 's comments have been address in my latest PRs.

Run it locally and it seems to work. The condition I'm using is `<j:when test="${anc.object == app and anc.object.displayName == 'Jenkins'}">`. Based on Daniel's suggestions I'm seeing now it seems the second part might not be necessary?

I was trying to debug the value of `anc.object`, but the Jelly file won't accept breakpoints when trying to "remote debug". Any hints on how to do it?

romenrg commented:
Simplified the conditional clause per @daniel-beck 's suggestion and tested it. It works great both in English and Spanish.

IMO this is ready now.

Thanks everyone, specially Daniel!

daniel-beck commented on core/src/main/resources/lib/layout/breadcrumbBar.jelly:
```diff
@@ -44,8 +44,16 @@ THE SOFTWARE.
-                      ${anc.object.displayName}
+                      <j:choose>
+                        <j:when test="${anc.object == app}">
+                          ${%dashboard}
```
Hint: If the English localization is simple enough and rather unlikely to change, it can be put here directly, saves having to create a "default" properties file.

Stylistic preference only.

romenrg replied:
If I understand correctly, your suggestion would be to use `${%Dashboard}` (title case) and remove the default properties, since it will use the property name as the value, right?

Can do that. Although when the i18n suggestion came along, I liked the idea of having a properties for the breadcrumbs, since I would expect it to eventually have more properties in it. I'll change it to title case and squash commits, but if it is decided to remove it, I'm OK with it.

daniel-beck commented on core/src/main/resources/lib/layout/breadcrumbBar.jelly:
```diff
@@ -44,8 +44,16 @@ THE SOFTWARE.
               <j:forEach var="anc" items="${request.ancestors}">
                 <j:if test="${h.isModel(anc.object) and anc.prev.url!=anc.url}">
                   <li class="item">
+                    <!-- Workaround to set desired name to first breadcrumb (instead of Jenkins) -->
```
Not: Would move into the choose block, otherwise context is unclear. Or just remove altogether.

daniel-beck commented on core/src/main/resources/lib/layout/breadcrumbBar_es.properties:
```diff
@@ -0,0 +1,23 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+dashboard=Panel de control
```
Should this be title-cased? It is elsewhere:

https://github.com/jenkinsci/jenkins/blob/22aa2e6e766074d11249893e3f35e0b99e20d3d0/core/src/main/resources/hudson/logging/LogRecorderManager/sidepanel_es.properties#L23

fqueiruga replied:
I would keep it Dashboard TBH, Panel de Control is too long I believe

romenrg replied:
I had the same doubt, @fqueiruga, but I saw it was being used already in other places and since I was trying to test the new breadcrumb at least in another language, I added it.

I'll use title case per Daniel's suggestion, but no hard opinion if it is decided to remove it.

romenrg commented:
I have:

- Moved the comment to the choose block
- Applied title case to Spanish translation (as noted in the comments, if decision is made to remove it, I'm OK with it; I used since it was already translated in other places and also to test another language)
- Changed "Dashboard" property to upper/title case (as also noted in the comments, if decision is made to remove the default properties file, I'm OK with it; I kept it bc I liked the idea of having a properties for the breadcrumbs, since I would expect it to eventually have more properties in it)

Finally, I have squashed all commits into one, as per the label @daniel-beck added.

Thank you for the feedback!

daniel-beck commented:
> Finally, I have squashed all commits into one, as per the label @daniel-beck added.

That's an instruction to maintainers once a PR is ready to be merged. It's generally better to not squash manually, as that messes with features like "new/unread commits" and line comments. Doesn't matter here, but in more complex PRs it can waste a lot of time of reviewers.

daniel-beck commented on core/src/main/resources/lib/layout/breadcrumbBar.properties:
```diff
@@ -0,0 +1,23 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+Dashboard=Dashboard
```
This file can now be deleted (and created anew when there's an actual need).

daniel-beck reviewed (approved):
Re-:+1:

MRamonLeon reviewed (approved):
Just don't forget to remove the properties file, not the Spanish one ;-)

romenrg committed:
[JENKINS-60972] Default properties file for breadcrumb is unnecessary

romenrg commented:
I've deleted the default properties file. We should be good now.

timja commented:
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!
